File: Algorithms/15. 3 Sum/3 Sum.java
//Author: Tushar Jaiswal
//Creation Date: 10/03/2015

/*Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note:
Elements in a triplet (a,b,c) must be in non-descending order. (ie, a = b = c)
The solution set must not contain duplicate triplets.
    Input: [-1,0,1,2,-1,-4]
    Output: [[-1,0,1],[-1,-1,2]]
    
    Input: [0,-4,-1,-4,-2,-3,2,0,0]
    Output:[[-2,0,2],[0,0,0]]*/

public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        
        int target = 0;
        List<List<Integer>> triplets = new ArrayList<List<Integer>>();
        
        Arrays.sort(nums);
        
        for(int i=0; i<nums.length; i++)
        {
            int twoSum = target-nums[i];
            int j = i+1;
            int k = nums.length-1;
            
            if(i > 0 && nums[i] == nums[i-1]) 
            { continue; }
            
            while(j < k)
            {
                if(twoSum > nums[j] + nums[k])
                { j++; }
                else if(twoSum < nums[j] + nums[k])
                { k--; }
                else
                {
                    triplets.add(new ArrayList<Integer>(Arrays.asList(nums[i], nums[j], nums[k])));
                    
                    //Avoiding duplicates
                    j++;
                    k--;
                    while(j<k && nums[j]==nums[j-1])
                    { j++; }
                    while(j<k && nums[k]==nums[k+1])
                    { k--; }
                }
            }
        }
        return triplets;
    }
}

File: Algorithms/15. 3 Sum/3Sum.cs
//Author: Tushar Jaiswal
//Creation Date: 02/08/2019

/*Given an array nums of n integers, are there elements a, b, c in nums such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.
Note: The solution set must not contain duplicate triplets.

Example:
Given array nums = [-1, 0, 1, 2, -1, -4],
A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]*/

public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        int target = 0;
        Array.Sort(nums);
        IList<IList<int>> result = new List<IList<int>>();

        for(int i = 0; i < nums.Length - 2; i++)
        {
            int twoSum = target - nums[i];
            int left = i + 1;
            int right = nums.Length - 1;

            if(i > 0 && nums[i - 1] == nums[i])
            { continue; }

            while(left < right)
            {
                int sum = nums[left] + nums[right];
                if(sum == twoSum)
                {
                    result.Add(new List<int>(new int[]{nums[i], nums[left], nums[right]}));
                    do
                    {
                        left++;
                    }while(left < right && nums[left] == nums[left - 1]);
                    do
                    {
                        right--;
                    }while(left < right && nums[right] == nums[right + 1]);
                }
                else if(sum < twoSum)
                {
                    left++;
                }
                else
                {
                    right--;
                }
            }
        }
        return result;
    }
}
